Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -1978,8 +1982,121 @@ - (void)handleCallTap:(TSCall *)call
     [self presentViewController:alertController animated:YES completion:nil];
 }
 
+#pragma mark - MessageActionsDelegate
+
+- (void)messageActionsShowDetailsForItem:(ConversationViewItem *)conversationViewItem
+{
+    [self showDetailViewForViewItem:conversationViewItem];
+}
+
+- (void)messageActionsReplyToItem:(ConversationViewItem *)conversationViewItem
+{
+    [self populateReplyForViewItem:conversationViewItem];
+}
+
+#pragma mark - MenuActionsViewControllerDelegate
+
+- (void)menuActionsDidHide:(MenuActionsViewController *)menuActionsViewController
+{
+    [[OWSWindowManager sharedManager] hideMenuActionsWindow];
+
+    [self updateShouldObserveDBModifications];
+}
+
+- (void)menuActions:(MenuActionsViewController *)menuActionsViewController
+    isPresentingWithVerticalFocusChange:(CGFloat)verticalChange
+{
+    UIEdgeInsets oldInset = self.collectionView.contentInset;
+    CGPoint oldOffset = self.collectionView.contentOffset;
+
+    UIEdgeInsets newInset = oldInset;
+    CGPoint newOffset = oldOffset;
+
+    // In case the message is at the very top or bottom edge of the conversation we have to have these additional
+    // insets to be sure we can sufficiently scroll the contentOffset.
+    newInset.top += verticalChange;
+    newInset.bottom -= verticalChange;
+    newOffset.y -= verticalChange;
+
+    DDLogDebug(@"%@ in %s verticalChange: %f, insets: %@ -> %@",
+        self.logTag,
+        __PRETTY_FUNCTION__,
+        verticalChange,
+        NSStringFromUIEdgeInsets(oldInset),
+        NSStringFromUIEdgeInsets(newInset));
+
+    // Because we're in the context of the frame-changing animation, these adjustments should happen
+    // in lockstep with the messageActions frame change.
+    self.collectionView.contentOffset = newOffset;
+    self.collectionView.contentInset = newInset;
```

Comment:
How do we ensure that `contentOffset` and `contentInset` are restored to their default values after dismissal?

Reply:
Similarly, will `contentOffset` and `contentInset` be restored properly if something other than the user initiates dismissal?  e.g. an incoming call or app becoming inactive?

Reply:
(I solved this the same as above)

> To solve, I took a slightly different approach using the delegate, rather than a notification, which I think is a little easier to get right with it's explicit sender/receiver.